core/primitives/src/stateless_validation/state_witness.rs

```diff
@@ -146,25 +173,50 @@ pub struct ChunkStateWitness {
     /// After these are applied as well, we should arrive at the pre-state-root
     /// of the chunk that this witness is for.
     pub implicit_transitions: Vec<ChunkStateTransition>,
-    #[deprecated(
-        note = "Was used for protocol versions without relaxed chunk validation which is not supported anymore."
-    )]
-    pub _deprecated_new_transactions: Vec<SignedTransaction>,
-    /// The following two field is deprecated and should not be use anymore
-    #[deprecated(
-        note = "Was used for protocol versions without relaxed chunk validation which is not supported anymore."
-    )]
-    pub _deprecated_new_transactions_validation_state: PartialState,
-    // TODO(stateless_validation): Deprecate this field in the next version of the state witness.
-    signature_differentiator: SignatureDifferentiator,
+}
+
+// Note: After deprecation of VersionedStateWitness protocol feature, we can remove this
+impl BorshSerialize for ChunkStateWitness {
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        match self {
+            ChunkStateWitness::V1(witness) => BorshSerialize::serialize(&witness, writer),
+            ChunkStateWitness::V2(witness) => {
+                BorshSerialize::serialize(&1_u8, writer)?;
+                BorshSerialize::serialize(&witness, writer)
+            }
+        }
+    }
+}
+
+// Note: After deprecation of VersionedStateWitness protocol feature, we can remove this
+impl BorshDeserialize for ChunkStateWitness {
+    /// Deserialize based on the first and second bytes of the stream. For V0, we do backward compatible deserialization by deserializing
+    /// the entire stream into V0. For V1, we consume the first byte and then deserialize the rest.
+    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
```

Comment:
The `signature_differentiator: SignatureDifferentiator` seems to already serve the purpose to differentiate between signed messages. I also don't see a reason to remove it because safety of interpreting signed bytes is a better thing. There were some discussions about it:

[#nearone/private > signature differentiators @ 💬](https://near.zulipchat.com/#narrow/channel/308695-nearone.2Fprivate/topic/signature.20differentiators/near/497989626)
There is also a NEP https://github.com/near/NEPs/pull/483. Doesn't seem active though.

Reply:
Umm, I think that's not the intention of the custom deserialize function here. Our current witness struct `ChunkStateWitnessV1` serializes as `borsh_serialize(witness)`. When we shift to an enum variant of `ChunkStateWitness`, the variant `ChunkStateWitness::V1(witness)` would directly serialize as `0x00 || borsh_serialize(witness)`, however for backward compatability, we would like to maintain the serialization of V1 struct as `borsh_serialize(witness)`, without the `0` prefix from the enum.

Going forward `ChunkStateWitness::V2(witness)` can be serialized as `0x01 || borsh_serialize(witness)`, and since chunk state witness isn't persisted beyond the GC period, we can deprecate V1 in a future release.

Reply:
Also, I was under the impression, the signature differentiator was mainly used for network messages? We had originally started with the `ChunkStateWitness` being the original message, but once we got partial witness in place, the signature differentiator is not a part of `PartialEncodedStateWitness`

Reply:
Ah, right. I forgot about that.

But something about the whole logic still looks confusing.

We don't distribute `ChunkStateWitness` by the network.
Instead, we split it into byte subslices and send them as `PartialEncodedStateWitness`es.
But doesn't `PartialEncodedStateWitness` already contain information which is enough to choose the right `ChunkStateWitness` version? As it has epoch_id, we know protocol version of state witness, so we know which version we should decode. Seems cleaner than looking into the raw bytes.

Reply:
That's a great idea! Let me try that.